server/SessionService.java

```diff
@@ -100,7 +100,7 @@ private void cancelIdleTimeout() {
     }
 
     private synchronized void mayStartIdleTimeout() {
-        if (isOpen() && transactionServices.isEmpty()) {
+        if (isOpen() && transactionServices.isEmpty() && idleTimeoutTask == null) {
```

Comment:
Is this completely thread-safe?

In theory, it is still possible for multiple concurrent invocations of the `mayStartIdleTimeout` to evaluate the if condition to `true` and then execute the if body, right?

Maybe adding a lock, or wrapping `idleTimeoutTask` within an `AtomicRef` solution would be the 100% foolproof solution?

Reply:
The methpd is `synchronized`, which means entering the method acquires a lock on the object and returning from it releases the lock. We're a-okay.